### DISCUSSION TITLE
How do I detect if a subscription to a collection has failed?

### DISCUSSION THREAD

**Discussion by deucks** [pocketbase/pocketbase#5624] [Q&A]
I'm trying to detect if a subscription to a collection has failed and then resubscribe if it has. 

I am attempting something like this 

```
``   const subscribeWithRetry = async (
      collectionName: string,
      callback: (e: any) => void,
      retryDelay: number = 5000
    ) => {
      try {
        await this.pb.collection(collectionName).subscribe("*", callback);
        console.log(`Subscribed to ${collectionName} collection.`);
      } catch (error) {
        console.error(
          `Subscription to ${collectionName} collection failed. Retrying in ${
            retryDelay / 1000
          } seconds...`,
          error
        );
        setTimeout(
          () => subscribeWithRetry(collectionName, callback, retryDelay),
          retryDelay
        );
      }
    };``
    
```
    
This however does not work, is there a way to listen for when it fails/unsubscribes?

**Comment by ganigeorgiev** [comment]
It is not clear what is the expected behavior you are looking for.

The subscribe call will throw usually if:

- cannot establish SSE connection with the server
- the underlying clientId submitted with `POST /api/realtime` is missing

It won't throw if the user doesn't have access the collection, or even technically if the collection name is missing because we allow custom topics. `pb.collection(collectionName).subscribe("*", ...)` is a convenient record wrapper for the more generic `pb.realtime.subscribe(collectionName + "/*", ...)`.
The access to the topic is determined based on each individual record and on the list or view API rules of the collection, but this doesn't throw and the clients that don't have access to the record change simply don't receive notification for it (note that the access to a record can change during runtime, for example an admin updating the API rules after a client has already subscribed).

To simulate `subscribe` failure you can create plain `pb_public/index.html` file:

```html
<!DOCTYPE html />
<html>
<head>
    <meta charset="utf-8" />
    <title>Test</title>
</head>
<body>
    <script type="module">
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.es.mjs"

        const pb = new PocketBase("invalid"); // invalid or inaccessible url should cause the subscribe to fail

        try {
            await pb.collection("example").subscribe("*", (e) => {
                console.log("change", e)
            })
        } catch (err) {
            console.log("Subscribe failed:", err)
        }
    </script>
</body>
</html>
```
Open your browser, navigate to [http://127.0.0.1:8090](http://127.0.0.1:8090) and check the browser console.

**Reply by ganigeorgiev** [reply]
_Additionally, I don't understand exactly what is the purpose of the manual retry but note that once a realtime connection is established the SDKs maintain their own autoreconnect in case the server drops the connection._

**Reply by deucks** [reply]
Thanks for taking the time to reply. I was hoping to get a callback for when the realtime is dropped. My use case requires the collection to be subscribed to continuously and if it does drop it should retry connecting indefinitely. This includes cases where the PB instance may be down. 
I have had issues where when connected for over 48+ hours, it drops the connection and doesn't receive any further realtime updates. (for example, any changes in the collection are no longer received, and I have to manually restart the application for it to resubscribe).  I want a way to verify that the web socket connection is valid.

Is there a way to listen to the websocket connection?

**Reply by ganigeorgiev** [reply]
There is no WebSocket, the realtime subscriptions are based on [Server-sent events (SSE)](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events).

I still don't understand what is the underlying issue. If the realtime connection interrupts, the SDKs should try to reconnect in the background indefinitely as long as there is at least 1 subscription.

If you want to attach a callback every time when a realtime connection is established, you can listen to the `PB_CONNECT` topic (aka. the first message that is always send by the server and that contains the `clientId`), for example:

```js
pb.realtime.subscribe("PB_CONNECT", (e) { ... })
```

Your problem is probably caused by something else related but without a minimal code sample illustrating the issue I don't think I can help you.

**Reply by deucks** [reply]
Thank you for the clarification. My issue is with the subscribe method, when it has been subscribed for over 48+ hours. It will unsubscribe and not recover, when it has no reason not to. There is also no callback when it has been disconnected, so I don't actually know when it has stopped.

I can provide some code, but I am using the standard subscribe method. Not sure if it'll help much.

**Reply by ganigeorgiev** [reply]
> It will unsubscribe and not recover, when it has no reason not to.

This doesn't really make sense to me.

If `onDisconnect` or other similar helper method to indicate dropped connection is what you need, then I would consider it with the PocketBase v0.23.0 release and its related compatible SDKs versions (there are no ETAs).

But again, I think what you are experiencing is a side-effect of something else and I'd advice to try to recreate a minimal reproducible repo in case you thing that that there is a bug with PocketBase or the SDKs.

**Reply by deucks** [reply]
Yes that would be really helpful, a `onDisconnect ` method would solve my problems. 

I'll try to reproduce the issue I am having, but it is very difficult as it produces no logs.
Do you think it could be GC causing the issue?

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here - with JS SDK v0.23.0 and Dart v0.20.0 there is now optional `pb.realtime.onDisconnect` hook function, but as mentioned above the SDKs perform autoreconnect on their own and this function shouldn't be needed for such cases. It could be useful though if you want to apply some special behavior after unsubscribing from all connections or to detect network/server errors._
